### PR TITLE
Use fsnotify lib to reload config

### PR DESCRIPTION
**Pull request by hongkailiu**
Alternative implementation of config reloading via [fsnotify](https://github.com/fsnotify/fsnotify)

* pros: our code is simpler
* cons: introducing a new dep.

### CODE REVIEW

File: pkg/controller/config/agent.go
Comment (petr-muller) [inline pkg/controller/config/agent.go]: IMO then it does not make sense to continue.

File: pkg/controller/config/agent.go
Comment (petr-muller) [inline pkg/controller/config/agent.go]: huh, TIL, I guess :/

File: pkg/controller/config/agent.go
Comment (petr-muller) [inline pkg/controller/config/agent.go]: this looks excessive, we'll log the event itself above

File: pkg/controller/config/agent.go
Comment (petr-muller) [inline pkg/controller/config/agent.go]: personally I would avoid doing the deep equality check just for the sake of logging, but I understand the positive side of that

File: pkg/controller/config/agent.go
Comment (hongkailiu) [reply pkg/controller/config/agent.go]: Done. `Fital` now.

File: pkg/controller/config/agent.go
Comment (hongkailiu) [reply pkg/controller/config/agent.go]: Yes. New to me too.
Copied from API example.
https://github.com/fsnotify/fsnotify/blob/master/example_test.go

File: pkg/controller/config/agent.go
Comment (hongkailiu) [reply pkg/controller/config/agent.go]: Removed the one above.

File: pkg/controller/config/agent.go
Comment (hongkailiu) [reply pkg/controller/config/agent.go]: Removed.
One of my bad choices.
`grep "modified file"` in the log should be good enough.

**Review by petr-muller — APPROVED** [review]
/lgtm
/approve
/hold

Holding for @droslean and  @stevekuznetsov 's thougts on new dep.

**Comment by droslean** [comment]
Personally, I think using `fsnotfy` is better, and I can live with this extra library. 
LGTM from my side.

@stevekuznetsov it's your call.

**Comment by bbguimaraes** [comment]
Have you tested this with an actual `ConfigMap` mount? Kubernetes creates a non-trivial directory structure, I think just watching the single file for writes using `inotify` is not sufficient:

```
$ oc -n ci exec secret-mirroring-controller-91-hfhgp -- ls -laR /etc/config/
/etc/config/:
total 0
drwxrwsrwx. 3 root 1000080000 79 Aug  2 18:16 .
drwxr-xr-x. 1 root root       20 Aug  2 15:52 ..
drwxr-sr-x. 2 root 1000080000 26 Aug  2 18:16 ..2019_08_02_18_16_30.006116574
lrwxrwxrwx. 1 root 1000080000 31 Aug  2 18:16 ..data -> ..2019_08_02_18_16_30.006116574
lrwxrwxrwx. 1 root root       19 Aug  2 15:52 mapping.yaml -> ..data/mapping.yaml

/etc/config/..2019_08_02_18_16_30.006116574:
total 4
drwxr-sr-x. 2 root 1000080000   26 Aug  2 18:16 .
drwxrwsrwx. 3 root 1000080000   79 Aug  2 18:16 ..
-rw-r--r--. 1 root 1000080000 1519 Aug  2 18:16 mapping.yaml
```

I suppose the update procedure is analogous to:

```
mkdir /etc/config/..newdir
oc get cm secret-mirroring -o jsonpath='{.data.mapping\.yaml}' > /etc/config/..newdir/mapping.yaml
ln -nsf /etc/config/..newdir /etc/config/..data
```

which would not result in a write to `/etc/config/mapping.yaml`.

**Comment by hongkailiu** [comment]
@bbguimaraes Let me test with a pod. Thx.

**Comment by hongkailiu** [comment]
@bbguimaraes you are right.
I cannot make the test (defined in https://github.com/openshift/ci-secret-mirroring-controller/pull/11) work even if i watch the `dir` where config is.

I will try more today.

**Comment by stevekuznetsov** [comment]
What's wrong with the config or secret agent impl from Prow? Why do we need something different?

**Comment by hongkailiu** [comment]
@bbguimaraes 
Got a working version.

But closing the issue because of https://coreos.slack.com/archives/GB7NB0CUC/p1565020515482900?thread_ts=1565020366.482400&cid=GB7NB0CUC

**Pull request closed without merging**
